### PR TITLE
feat: Introduce Pgvector in benchmarks

### PR DESCRIPTION
**Pull request by KShivendu**
Introduce Pgvector in benchmarks

### CODE REVIEW

File: engine/clients/pgvector/parser.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        gt: Optional[FieldValue],
+        lte: Optional[FieldValue],
+        gte: Optional[FieldValue],
+    ) -> Any:
+        raise NotImplementedError()
+
+    def build_geo_filter(
+        self, field_name: str, lat: float, lon: float, radius: float
+    ) -> Any:
+        raise NotImplementedError()
```
Comment (timvisee) [inline engine/clients/pgvector/parser.py:28]: Why are all of these marked as not yet implemented at this time?

File: engine/clients/pgvector/configure.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+        self.cursor.execute(
+            f"""CREATE TABLE items (
+                id SERIAL PRIMARY KEY,
+                embedding vector({dataset.config.vector_size}) NOT NULL
+            );"""
+        )
+        self.cursor.execute(
+            "ALTER TABLE items ALTER COLUMN embedding SET STORAGE PLAIN"
+        )
```
Comment (timvisee) [inline engine/clients/pgvector/configure.py:41]: If we can recreate here, do we actually need the `init.sql`?

**Review by timvisee — APPROVED** [review]
LGTM

Though please take this with a grain of salt, I've limited understanding of our benchmarking suite.

File: engine/clients/pgvector/parser.py (hunk 1 above)
Comment (KShivendu) [reply engine/clients/pgvector/parser.py:28]: We just want to update [Single Node Benchmarks](https://qdrant.tech/benchmarks/#single-node-speed-benchmark) for now. The datasets involved don't have any filters. It's all embeddings. Hence these aren't required.

File: engine/clients/pgvector/configure.py (hunk 2 above)
Comment (KShivendu) [reply engine/clients/pgvector/configure.py:41]: Good point. I'll drop `init.sql`

File: engine/clients/pgvector/configure.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+        if dataset.config.distance == Distance.COSINE:
+            hnsw_distance_type = "vector_cosine_ops"
```
Comment (joein) [inline engine/clients/pgvector/configure.py:44]: hnsw distance type should be taken from `self.DISTANCE_MAPPING`

File: engine/clients/pgvector/search.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+from engine.clients.pgvector.parser import PgVectorConditionParser
+
+
+class PgVectorSearcher(BaseSearcher):
+    search_params = {}
+    cursor = None
+    parser = PgVectorConditionParser()
+
+    @classmethod
+    def get_mp_start_method(cls):
```
Comment (joein) [inline engine/clients/pgvector/search.py:21]: just to be sure:
we use this when a database client does not work with "fork", have you checked if it is indeed required to have "forkserver" or "spawn" for pgvector here?

File: engine/clients/pgvector/search.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        return "forkserver" if "forkserver" in mp.get_all_start_methods() else "spawn"
+
+    @classmethod
+    def init_client(cls, host, distance, connection_params: dict, search_params: dict):
+        cls.conn = psycopg2.connect(**get_db_config(host))
```
Comment (joein) [inline engine/clients/pgvector/search.py:26]: `cls.conn` should be defined as a class member (check how it is implemented in QdrantSearcher)

File: engine/clients/pgvector/search.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        register_vector(cls.conn)
+        cls.cur = cls.conn.cursor(cursor_factory=RealDictCursor)
```
Comment (joein) [inline engine/clients/pgvector/search.py:28]: `cls.cur` - same as with `cls.conn`

File: engine/clients/pgvector/search.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        cls.distance = distance
```
Comment (joein) [inline engine/clients/pgvector/search.py:29]: `cls.distance` - same as with `cls.conn` and `cls.cur`

File: engine/clients/pgvector/search.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        cls.connection_params = connection_params
+        cls.search_params = search_params["search_params"]
+
+    @classmethod
+    def search_one(cls, vector, meta_conditions, top) -> List[Tuple[int, float]]:
+        cls.cur.execute("SET hnsw.ef_search = %s", (cls.search_params["hnsw_ef"],))
+
+        if cls.distance == Distance.COSINE:
+            QUERY = f"SELECT id, embedding <=> %s AS _score FROM items ORDER BY _score LIMIT {top};"
```
Comment (joein) [inline engine/clients/pgvector/search.py:38]: variable name should be lower case, it is not a constant

File: engine/clients/pgvector/upload.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
+from io import BytesIO, StringIO
```
Comment (joein) [inline engine/clients/pgvector/upload.py:1]: stringio - redundant import

File: engine/clients/pgvector/upload.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+from engine.clients.pgvector.config import get_db_config
+
+
+class PgVectorUploader(BaseUploader):
+    client = None
+    upload_params = {}
+
+    @classmethod
+    def init_client(cls, host, distance, connection_params, upload_params):
+        cls.conn = psycopg2.connect(**get_db_config(host))
```
Comment (joein) [inline engine/clients/pgvector/upload.py:17]: same as in searcher with `cls.conn`, `cls.curr` and `connection_params`

**Review by joein — CHANGES_REQUESTED** [review]
please look at my comments above

File: engine/clients/pgvector/configure.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        elif dataset.config.distance == Distance.L2:
+            hnsw_distance_type = "vector_l2_ops"
+        else:
+            raise NotImplementedError(f"Unsupported distance metric: {self.metric}")
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply engine/clients/pgvector/configure.py:48]: Fixed. Using `dataset.config.distance`

File: engine/clients/pgvector/configure.py (hunk 3 above)
Comment (KShivendu) [reply engine/clients/pgvector/configure.py:44]: Done. Thanks!

File: engine/clients/pgvector/search.py (hunk 5 above)
Comment (KShivendu) [reply engine/clients/pgvector/search.py:26]: Done

File: engine/clients/pgvector/search.py (hunk 6 above)
Comment (KShivendu) [reply engine/clients/pgvector/search.py:28]: Done. Created a class variable

File: engine/clients/pgvector/search.py (hunk 8 above)
Comment (KShivendu) [reply engine/clients/pgvector/search.py:38]: Fixed.

File: engine/clients/pgvector/search.py (hunk 4 above)
Comment (KShivendu) [reply engine/clients/pgvector/search.py:21]: Will check. Thanks!

File: engine/clients/pgvector/search.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    def get_mp_start_method(cls):
+        return "forkserver" if "forkserver" in mp.get_all_start_methods() else "spawn"
+
+    @classmethod
+    def init_client(cls, host, distance, connection_params: dict, search_params: dict):
+        cls.conn = psycopg2.connect(**get_db_config(host))
+        register_vector(cls.conn)
+        cls.cur = cls.conn.cursor(cursor_factory=RealDictCursor)
+        cls.distance = distance
+        cls.connection_params = connection_params
```
[the remark this answers is not in the document]
Comment (KShivendu) [reply engine/clients/pgvector/search.py:30]: Indeed yes. Made some changes in get_db_config and now we can just do 
```python
psycopg2.connect(**get_db_config(host, connection_params))
```

Updated all the occurences.

File: engine/clients/pgvector/upload.py (hunk 9 above)
Comment (KShivendu) [reply engine/clients/pgvector/upload.py:1]: Removed.

File: engine/clients/pgvector/parser.py (hunk 1 above)
Comment (KShivendu) [reply engine/clients/pgvector/parser.py]: Just implemented all the functions in `PgVectorConditionParser`

File: engine/clients/pgvector/search.py (hunk 4 above)
Comment (KShivendu) [reply engine/clients/pgvector/search.py:23]: It wasn't required. Removed it.

**Comment by KShivendu** [comment]
Closing in favour of https://github.com/qdrant/vector-db-benchmark/pull/91

**Pull request closed without merging**